nomad/structs/structs.go

```diff
@@ -9998,6 +10012,85 @@ func (a *Allocation) AllocationDiff() *AllocationDiff {
 	return (*AllocationDiff)(a)
 }
 
+// Expired determines whether an allocation has exceeded its MaxClientDisonnect
+// duration relative to the passed time stamp.
+func (a *Allocation) Expired(now time.Time) bool {
+	// If alloc is not Unknown it cannot be expired.
+	if a.ClientStatus != AllocClientStatusUnknown {
+		return false
+	}
+
+	lastUnknown := a.LastUnknown()
+	if lastUnknown.IsZero() {
+		return false
+	}
+
+	return a.expired(now, lastUnknown)
+}
+
+func (a *Allocation) expired(now, lastUnknown time.Time) bool {
```

Comment:
It is trying to avoid multiple calls to `LastUnknown` which iterates over `AllocStates`.  `Expired` and `Reconnected` both need to do this. It's also trying a way of checking the expiration without checking the `ClientStatus`. You just spotted a bug in Reconnected though. It can't call `Expired` and be valid.  I'll rework this.

Reply:
I think this was a refactoring that went sideways.  I've made it much simpler. Sorry for the oops. I'm leaving this comment open so you can review it in the next push.

Reply:
Much nicer!